Sources/NIO/ChannelOption.swift

```diff
@@ -225,9 +255,21 @@ extension ChannelOptions {
 
 /// Provides `ChannelOption`s to be used with a `Channel`, `Bootstrap` or `ServerBootstrap`.
 public struct ChannelOptions {
+    #if !os(Windows)
+        @available(*, deprecated, renamed: "ChannelOptions.socketOption(_:)")
+        public static let socket = { (level: SocketOptionLevel, name: SocketOptionName) -> Types.SocketOption in
+            .init(level: NIOBSDSocket.OptionLevel(rawValue: CInt(level)), name: NIOBSDSocket.Option(rawValue: CInt(name)))
+        }
+    #endif
+
+    /// - seealso: `SocketOption`.
+    public static let socketOption = { (name: NIOBSDSocket.Option) -> Types.SocketOption in
```

Comment:
I'm wondering if we need a third one of these that does not default the `level` parameter so that folks have an escape hatch. Thoughts?

Reply:
Isn’t that what they have through the old API?  I agree that’s not as convenient.

Reply:
The old API uses different types, and is deprecated. But there is more than just `SOL_SOCKET` and `IPPROTO_TCP`. For example, there's `IPPROTO_IP`. Rather than try to cover them all in one go, we should just provide a place where users can provide both the level and name.

Reply:
@Lukasa I don't think we should / can deprecate the old API. I thought the new stuff is a shorthand for `SOL_SOCKET`. Before we deprecate the old API (which will break every NIO app), we should have an actually _nice_ one like https://github.com/apple/swift-nio/pull/1260 has started to do. Once we have migrated 99% of our users to "nice" `ChannelOptions` that are also usable for Network.framework, then we can start deprecating `.socket(.., ...)` I think

Reply:
@weissi You may want to re-review this PR then: it adds a completely new namespace for socket options and fundamentally changes their types.

Reply:
I thought that they old API just was unavailable on Windows but not marked as deprecated.

Reply:
agree with @compnerd

Reply:
@Lukasa the way I understand this patch is that it gives us a new, fully compatible (but any nicer) way of spelling `.socket(.init(SOL_SOCKET), *)` for BSD Sockets, nothing more or less. So it would be 100% compatible with `.socket(.init(SOL_SOCKET), ...)` on UNIX.

On Windows `.socket` would be entirely unavailable and for now we only offer `.socketOption`.

As a separate patch (not done by @compnerd probably), we'd actually offer nice APIs that don't contain the word "socket" for example and look actually "swifty".

@compnerd is that a fair summary?

Reply:
The patch as it currently stands makes the old API unavailable on Windows and deprecated everywhere else:

https://github.com/apple/swift-nio/blob/0161bd76b2a6eedebc955738deb0478d0fcbd1af/Sources/NIO/ChannelOption.swift#L95-L106

Moreover, to achieve @compnerd's goals we _must_ deprecate the old API. If we do not deprecate the old API, nothing will work on Windows because no-one will move off the current API. While we can move all the NIO examples, the wide range of third-party projects will not move because nothing will tell them that they could move, forcing @compnerd to chase them all. I do not think @compnerd would consider that a success.

Reply:
Sorry, wrong snippet, I wanted this one:

https://github.com/apple/swift-nio/blob/0161bd76b2a6eedebc955738deb0478d0fcbd1af/Sources/NIO/ChannelOption.swift#L258-L263

Reply:
Okay, I'm not happy with the deprecation. We can totally start migrating all the examples to using `.socketOption` instead of `.socket(SOL_SOCKET, ...)` but I don't think we can deprecate.

The reason is: We already did some changes to the socket options, now we're thinking about more changes but what I really want is an _actually nice_ API. So I think it's not cool to deprecate whatever was the last best mechanism all too often.

Reply:
I think that waiting on the subsequent patch that @weissi is talking about for the deprecation to follow in short order is acceptable.  I'm okay with removing the `@available` on `socket` for the time being with the assumption that there is ongoing work to deprecate it.

Reply:
Great, works for me. @weissi is anyone actually _working_ on that patch?

Reply:
ping @Davidde94

Reply:
Yeh I need to go back to the PR and get it merged. I can try and get this done tomorrow

Reply:
Okay, I think that we should try to get this merged in the mean time - it will allow me to start getting a whole slew of other changes drained.

Reply:
@compnerd totally agreed, I think this is just on hold because it's still breaking public API (says `swift-api-digester`)